Kernel/Devices/USB/UHCIDescriptorTypes.h

```diff
@@ -0,0 +1,285 @@
+
+//
+// Queue Head
+//
+// Description here please!
+//
+struct alignas(16) QueueHead {
+    enum class LinkPointerBits : u32 {
+        Terminate = 1,
+        QHSelect = 2,
+    };
+
+    QueueHead() = delete;
+    QueueHead(u32 paddr)
+        : m_paddr(paddr)
+    {
+    }
+    //~QueueHead() = delete; // Prevent anything except placement new on this object
+
+    u32 link_ptr() const { return m_link_ptr; }
+    u32 element_link_ptr() const { return m_element_link_ptr; }
+    u32 paddr() const { return m_paddr; }
+    bool in_use() const { return m_in_use; }
+
+    void set_in_use(bool in_use) { m_in_use = in_use; }
+    void set_link_ptr(u32 val) { m_link_ptr = val; }
+
+    // FIXME: For the love of God, use AK SMART POINTERS PLEASE!!
+    QueueHead* next_qh() { return m_next_qh; }
+    const QueueHead* next_qh() const { return m_next_qh; }
+    void set_next_qh(QueueHead* qh) { m_next_qh = qh; }
+
+    QueueHead* prev_qh() { return m_prev_qh; }
+    const QueueHead* prev_qh() const { return m_prev_qh; }
+    void set_previous_qh(QueueHead* qh)
+    {
+        m_prev_qh = qh;
+    }
+
+    void link_next_queue_head(QueueHead* qh)
+    {
+        m_link_ptr = qh->paddr();
+        m_link_ptr |= static_cast<u32>(LinkPointerBits::QHSelect);
+        set_next_qh(qh);
+    }
+
+    void terminate_with_stray_descriptor(TransferDescriptor* td)
+    {
+        m_link_ptr = td->paddr();
+        m_link_ptr |= static_cast<u32>(LinkPointerBits::Terminate);
+    }
+
+    // TODO: Should we pass in an array or vector of TDs instead????
+    void attach_transfer_descriptor_chain(TransferDescriptor* td)
+    {
+        m_first_td = td;
+        m_element_link_ptr = td->paddr();
+    }
+
+    void terminate() { m_link_ptr |= static_cast<u32>(LinkPointerBits::Terminate); }
+
+    void terminate_element_link_ptr()
+    {
+        m_element_link_ptr = static_cast<u32>(LinkPointerBits::Terminate);
+    }
+
+    // Clean the chain of transfer descriptors
+    void clean_chain()
+    {
+        // TODO
+    }
+
+    void print()
+    {
+        kprintf("UHCI: QH(%p) @ 0x%08x: link_ptr=0x%08x, element_link_ptr=0x%08x\n", this, m_paddr, m_link_ptr, m_element_link_ptr);
+    }
+
+private:
+    u32 m_link_ptr { 0 };                  // Pointer to the next horizontal object that the controller will execute after this one
+    volatile u32 m_element_link_ptr { 0 }; // Pointer to the first data object in the queue (can be modified by hw)
+
+    // These values will be ignored by the controller, but we can use them for configuration/bookkeeping
+    // Any addresses besides `paddr` are assumed virtual and can be dereferenced
+    u32 m_paddr { 0 };                          // Physical address where this QueueHead is located
+    QueueHead* m_next_qh { nullptr };           // Next QH
+    QueueHead* m_prev_qh { nullptr };           // Previous QH
+    TransferDescriptor* m_first_td { nullptr }; // Pointer to first TD
```

Comment:
I feel like these could be `OwnPtr<QueueHead>` instead?

Reply:
I don't see you calling `delete` on them anywhere, so would be the purpose of wrapping them in an owning pointer?

Reply:
I think this comes more from the mentality of "don't use raw pointers!!!" even though in this situation they're justified.